doc/developer/scan.txt

```diff
@@ -135,19 +60,252 @@ files, and section of the code they deal with, are :
+Non-recurring output (nitsot)                                *No corresponding outer input*                         *No corresponding inner input*                              Output value at timestep `t`                                 Concatenation of the values of the output at all timestep  `outputs_info`
+Singly-recurrent output (sitsot)                             Initial value (value at timestep `-1`)                 Output value at previous timestep (`t-1`)                   Output value at timestep `t`                                 Concatenation of the values of the output at all timestep  `outputs_info`
+Multiply-recurrent output (mitsot)                           Initial values for the required timesteps where `t<0`  Output value at previous required timesteps                 Output value at timestep `t`                                 Concatenation of the values of the output at all timestep  `outputs_info`
+Multiply-recurrent multiple outputs (mitmot)                 Initial values for the required timesteps where `t<0`  Output value at previous required timesteps                 Output values for current and multiple future timesteps      Concatenation of the values of the output at all timestep  *No corresponding argument*
+===========================================================  =====================================================  ==========================================================  ===========================================================  =========================================================  ======================================================
+
+
+Optimizations
+=============
+
+remove_constants_and_unused_inputs_scan
+---------------------------------------
+
+This optimization serves two purposes, The first is to remove a Scan Op's
+unused inputs. The second it to take a Scan Op's constant inputs and remove
+them, instead injecting the constants directly into the graph or the Scan
+Op's inner function. This will allow constant folding to happen inside the
+inner function.
+
+
+PushOutNonSeqScan
+-----------------
+
+This optimizations pushes, out of Scan's inner function and into the outer
+function, computation that depends only on non-sequence inputs. Such
+computation ends up being done every iteration on the same values so moving
+it to the outer function to be executed only once, before the Scan Op,
+reduces the amount of computation that needs to be performed.
+
+
+PushOutSeqScan
+--------------
+
+This optimization resembles PushOutNonSeqScan but it tries to push, out of
+the inner function, the computation that only relies on sequence and
+non-sequence inputs. The idea behing this optimization is that, when it is
+possible to do so, it is generally more computationnaly efficient to perform
+a single operation on a large tensor rather then perform that same operation
+many times on many smaller tensors. However, this optimization can lead to
+increased memory usage.
```

Comment:
I don't know. I don't strongly lean one way or the other. On one hand it might but useful for them to know, on the other hand could this sort of additional info that relies on them knowing about specific other stuff get confusing for some people? If you have a strong preference, I will do it.